### PR TITLE
FiLM layer for TPU training

### PR DESCRIPTION
**Pull request by dizcology**
Customer asked whether the [feature wise linear modulation](https://arxiv.org/abs/1709.07871) layer can be trained on TPUs.  This PR implements a working prototype.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/302)

### CODE REVIEW

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    x = features['x']
+    modulation_data = features['modulation_data']
+    onehot_labels = tf.one_hot(labels, n_classes)
+
+    global_step = tf.train.get_global_step()
+
+    # In this sample we use dense layers for the modulation sub-network.
+    # Its output has shape (batch_size, 2 * n_film) since each FiLM layer has
+    # two parameters.
+    modulation_hidden = tf.layers.dense(modulation_data, 128, activation=tf.nn.relu)
```
Comment (gogasca) [inline tpu/templates/tpu_film_estimator/trainer.py]: tf.layers will be deprecated in TF 2.0, not sure if makes sense to move it to tf.keras ?

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    dataset = tf.data.Dataset.from_tensor_slices((x_tensor, y_tensor, modulation_data_tensor))
+
+    # TPUEstimator passes params when calling input_fn
+    batch_size = params.get('batch_size', 16)
+
+    dataset = dataset.repeat().shuffle(32).batch(batch_size, drop_remainder=True)
+
+    # TPUs need to know all dimensions when the graph is built
+    # Datasets know the batch size only when the graph is run
+    def set_shapes_and_format(x, y, modulation_data):
```
Comment (gogasca) [inline tpu/templates/tpu_film_estimator/trainer.py]: Can we add Documentation information to the function as:
```
def foo():
  """This is a foo function which does x.

   Args:
      x: (type) This an argument.

   Returns:
      A foo value.
  """
```

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        '--use-tpu',
+        action='store_true')
+    parser.add_argument(
+        '--tpu',
+        default=None)
+
+    args, _ = parser.parse_known_args()
+
+    # colab.research.google.com specific
+    import sys
```
Comment (gogasca) [inline tpu/templates/tpu_film_estimator/trainer.py]: imports should be at top level

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    if 'google.colab' in sys.modules:
+        import json
+        import os
+
+        # TODO(user): change this
+        args.model_dir = 'gs://your-gcs-bucket'
+
+        # When connected to the TPU runtime
+        if 'COLAB_TPU_ADDR' in os.environ:
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: Not at all.  The .py --> .ipynb tool is too lazy about this.  Let me fix it in a separate PR since that fix will touch several files not in the scope of this PR.

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        y.set_shape(y_shape)
+        modulation_data.set_shape(modulation_data_shape)
+
+        # Also format the dataset with a dict for features
+        features = {'x': x, 'modulation_data': modulation_data}
+        labels = y
+
+        return features, labels
+
+    dataset = dataset.map(set_shapes_and_format)
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: We can, but I feel that having these dataset operations in separate lines makes it clearer to the reader of the code.

File: tpu/templates/tpu_film_estimator/trainer.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import numpy as np
+import tensorflow as tf
+from tensorflow.contrib import summary
+
+n_classes = 10
```
[the remark this answers is not in the document]
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: done.

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 1 above)
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: done, changed the model building code to use tf.keras.layers

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 2 above)
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: this is an inner function that probably would have been a `lambda`, as opposed to a library function to be reused out of context.  i feel the preceeding comment clarifies its purpose well enough.  i added a docstring.

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 3 above)
Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.py]: let me also fix this in the separate PR to remove the Colab specific code from the .py files.

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gogasca) [reply tpu/templates/tpu_film_estimator/trainer.py]: LGTM

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gogasca) [reply tpu/templates/tpu_film_estimator/trainer.py]: Sounds good, let's keep original behavior

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 2 above)
Comment (gogasca) [reply tpu/templates/tpu_film_estimator/trainer.py]: I agree, comment is self-explanatory. Just would like to standardize our samples and use a common structure.

File: tpu/templates/tpu_film_estimator/trainer.py (hunk 3 above)
Comment (gogasca) [reply tpu/templates/tpu_film_estimator/trainer.py]: Sounds good

File: tpu/templates/tpu_film_estimator/trainer.ipynb [hunk 7]
```diff
@@ -0,0 +1,384 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Copyright 2018 Google LLC  \n",
```
Comment (gogasca) [inline tpu/templates/tpu_film_estimator/trainer.ipynb]: This is great work, I see that there is reference a paper, can we add some detail around this paper in layman terms? ML problem/Dataset ?
Example:
https://colab.sandbox.google.com/github/tensorflow/tpu/blob/master/tools/colab/bert_finetuning_with_cloud_tpus.ipynb#scrollTo=uu2dQ_TId-uH

Thanks

Comment (dizcology) [reply tpu/templates/tpu_film_estimator/trainer.ipynb]: added a little context on FiLM.

**Comment by dizcology** [comment]
merging this now, and I'll return with a separate PR to remove colab-specific code from the .py files.
